nav2_behavior_tree/plugins/decorator/truncate_path_node.cpp

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2018 Intel Corporation
+// Copyright (c) 2020 Francisco Martin Rico
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include <memory>
+#include <limits>
+
+#include "nav_msgs/msg/path.hpp"
+#include "geometry_msgs/msg/pose_stamped.hpp"
+#include "nav2_util/geometry_utils.hpp"
+#include "behaviortree_cpp_v3/decorator_node.h"
+
+#include "nav2_behavior_tree/plugins/decorator/truncate_path_node.hpp"
+
+namespace nav2_behavior_tree
+{
+
+TruncatePath::TruncatePath(
+  const std::string & name,
+  const BT::NodeConfiguration & conf)
+: BT::DecoratorNode(name, conf),
+  distance_(1.0)
+{
+  getInput("distance", distance_);
+}
+
+inline BT::NodeStatus TruncatePath::tick()
+{
+  setStatus(BT::NodeStatus::RUNNING);
+
+  BT::NodeStatus ret_status = child_node_->executeTick();
+
+  nav_msgs::msg::Path input_path;
+
+  getInput("input_path", input_path);
+
+  if (input_path.poses.empty()) {
+    setOutput("output_path", input_path);
+    return ret_status;
+  }
+
+  geometry_msgs::msg::PoseStamped final_pose = input_path.poses.back();
+
+  double distance_to_goal = nav2_util::geometry_utils::euclidean_distance(
+    input_path.poses.back(), final_pose);
+
+  while (distance_to_goal < distance_) {
```

Comment:
A potential problem with this loop. If the updated goal is within the target distance already, the path computed will not be empty but all the points in the computed path will be within the target distance. `input_path` will not be empty in that case and all poses will be popped out of the path which will cause a segmentation fault.

Reply:
That is a good point. Though maybe if new goal is inside distance, then you should consider the following path 'done' so having an empty output path might be what is intended. 

But for the general truncate path node, should we handle that situation differently?

Reply:
I think that changing the while condition to 

```
while (distance_to_goal < distance_ && input_path.poses.size() > 2) {
```

will work. At least, we keep the first point in the path. So, no movement, but the robot orientates to the final pose. What do you think?

Reply:
Yup, that should work